File: spark-doris-connector/spark-doris-connector-base/src/main/scala/org/apache/doris/spark/config/DorisConfig.scala
package org.apache.doris.spark.config

import org.apache.doris.spark.exception.OptionRequiredException
import org.apache.spark.SparkConf
import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap

import scala.collection.JavaConverters.mapAsScalaMapConverter
import scala.reflect.ClassTag

class DorisConfig private (var configOptions: CaseInsensitiveMap[String]) extends Serializable {

  private val DORIS_REQUEST_AUTH_USER = "doris.request.auth.user"

  private val DORIS_REQUEST_AUTH_PASSWORD = "doris.request.auth.password"

  private def this(options: Map[String, String]) = {
    this(CaseInsensitiveMap(options))
    configOptions = CaseInsensitiveMap(processOptions(options))
    checkOptions(configOptions)
  }

  private def processOptions(options: Map[String, String]): Map[String, String] = {
    val dottedOptions: Map[String, String] = options.map {
      case (k, v) =>
        if (k.startsWith("sink.properties.") || k.startsWith("doris.sink.properties.")) {
          (k, v)
        } else {
          (k.replace('_', '.'), v)
        }
    }
    val processedOptions: Map[String, String] = dottedOptions.filter(!_._1.startsWith("spark.")).map {
      case (k, v) =>
        if (k.startsWith("doris.")) (k, v)
        else ("doris." + k, v)
    }.map {
      case (DORIS_REQUEST_AUTH_USER, v) => (DorisOptions.DORIS_USER.name, v)
      case (DORIS_REQUEST_AUTH_PASSWORD, v) => (DorisOptions.DORIS_PASSWORD.name, v)
      case (k, v) => (k, v)
    }
    processedOptions
  }

  private def checkOptions(options: Map[String, String]): Unit = {

    // require(options.contains(DorisOptions.DORIS_FENODES.name), s"option [${DorisOptions.DORIS_FENODES.name}] is required")
    if (!options.contains(DorisOptions.DORIS_FENODES.name)) {
      throw new OptionRequiredException(DorisOptions.DORIS_FENODES.name)
    } else {
      val feNodes = options(DorisOptions.DORIS_FENODES.name)
      if (feNodes.isEmpty) {
        throw new IllegalArgumentException(s"option [${DorisOptions.DORIS_FENODES.name}] is empty")
      } else if (!feNodes.split(",").exists(_.contains(":"))) {
        throw new IllegalArgumentException(s"option [${DorisOptions.DORIS_FENODES.name}] is not in correct format, for example: host:port[,host2:port]")
      }
    }
    if (!options.contains(DorisOptions.DORIS_TABLE_IDENTIFIER.name)) {
      throw new OptionRequiredException(DorisOptions.DORIS_TABLE_IDENTIFIER.name)
    } else {
      val tableIdentifier = options(DorisOptions.DORIS_TABLE_IDENTIFIER.name)
      if (tableIdentifier.isEmpty) {
        throw new IllegalArgumentException(s"option [${DorisOptions.DORIS_TABLE_IDENTIFIER.name}] is empty")
      } else if (!tableIdentifier.contains(".")) {
        throw new IllegalArgumentException(s"option [${DorisOptions.DORIS_TABLE_IDENTIFIER.name}] is not in correct format, for example: db.table")
      }
    }
    if (!options.contains(DorisOptions.DORIS_USER.name)) {
      throw new OptionRequiredException(DorisOptions.DORIS_USER.name)
    } else if (options(DorisOptions.DORIS_USER.name).isEmpty) {
      throw new IllegalArgumentException(s"option [${DorisOptions.DORIS_USER.name}] is empty")
    }
    if (!options.contains(DorisOptions.DORIS_PASSWORD.name)) {
      throw new OptionRequiredException(DorisOptions.DORIS_PASSWORD.name)
    }
  }

  def contains(option: ConfigOption[_]): Boolean = {
    configOptions.contains(option.name)
  }

  def getValue[T](option: ConfigOption[T])(implicit classTag: ClassTag[T]): T = {
    (configOptions.contains(option.name), option.defaultValue.nonEmpty) match {
      case (true, _) =>
        val originVal = configOptions(option.name)
        classTag.runtimeClass match {
          case c if c == classOf[Int] => originVal.toInt.asInstanceOf[T]
          case c if c == classOf[Long] => originVal.toLong.asInstanceOf[T]
          case c if c == classOf[Double] => originVal.toDouble.asInstanceOf[T]
          case c if c == classOf[Boolean] => originVal.toBoolean.asInstanceOf[T]
          case c if c == classOf[String] => originVal.asInstanceOf[T]
          case _ => throw new IllegalArgumentException(s"Unsupported config value type: ${classTag.runtimeClass}")
        }
      case (false, true) => option.defaultValue.get
      case _ => throw new OptionRequiredException(option.name)
    }
  }

  def setProperty(option: ConfigOption[_], value: String): Unit = {
    configOptions += option.name -> value
    configOptions = CaseInsensitiveMap(processOptions(configOptions))
    checkOptions(configOptions)
  }

  def getSinkProperties: Map[String, String] = configOptions.filter(_._1.startsWith(DorisOptions.STREAM_LOAD_PROP_PREFIX))
    .map {
      case (k, v) => (k.substring(DorisOptions.STREAM_LOAD_PROP_PREFIX.length), v)
    }

  def merge(map: Map[String, String]): Unit = {
    configOptions ++= map
    configOptions = CaseInsensitiveMap(processOptions(configOptions))
    checkOptions(configOptions)
  }

  def toMap: Map[String, String] = configOptions.toMap

}

object DorisConfig {

  def fromSparkConf(sparkConf: SparkConf): DorisConfig = {
    new DorisConfig(sparkConf.getAll.toMap)
  }

  def fromSparkConf(sparkConf: SparkConf, options: Map[String, String]): DorisConfig = {
    new DorisConfig(sparkConf.getAll.toMap ++ options)
  }

  def fromMap(map: Map[String, String]): DorisConfig = {
    new DorisConfig(map)
  }

  def fromMap(map: java.util.Map[String, String]): DorisConfig = {
    new DorisConfig(map.asScala.toMap)
  }

}

File: spark-doris-connector/spark-doris-connector-base/src/main/java/org/apache/doris/spark/config/DorisConfig.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.spark.config;

import org.apache.commons.collections.MapUtils;
import org.apache.doris.spark.exception.OptionRequiredException;

import java.io.Serializable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class DorisConfig implements Serializable {

    private final String DORIS_REQUEST_AUTH_USER = "doris.request.auth.user";
    private final String DORIS_REQUEST_AUTH_PASSWORD = "doris.request.auth.password";

    private Map<String, String> configOptions;
    private boolean ignoreTableCheck;

    // only for test
    public DorisConfig() {
        configOptions = Collections.emptyMap();
    }

    private DorisConfig(Map<String, String> options, Boolean ignoreTableCheck) throws OptionRequiredException {
        this.configOptions = new HashMap<>(processOptions(options));
        this.ignoreTableCheck = ignoreTableCheck;
        checkOptions(this.configOptions);
    }

    private Map<String, String> processOptions(Map<String, String> options) {
        Map<String, String> dottedOptions = new HashMap<>();
        for (Map.Entry<String, String> entry : options.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            if (key.startsWith("sink.properties.") || key.startsWith("doris.sink.properties.")) {
                dottedOptions.put(key, value);
            } else {
                dottedOptions.put(key.replace('_', '.'), value);
            }
        }

        Map<String, String> processedOptions = new HashMap<>();
        for (Map.Entry<String, String> entry : dottedOptions.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            if (!key.startsWith("spark.")) {
                if (key.startsWith("doris.")) {
                    processedOptions.put(key, value);
                } else {
                    processedOptions.put("doris." + key, value);
                }
            }
        }
        if (processedOptions.containsKey(DORIS_REQUEST_AUTH_USER)) {
            processedOptions.put(DorisOptions.DORIS_USER.getName(), processedOptions.get(DORIS_REQUEST_AUTH_USER));
        }
        if (processedOptions.containsKey(DORIS_REQUEST_AUTH_PASSWORD)) {
            processedOptions.put(DorisOptions.DORIS_PASSWORD.getName(), processedOptions.remove(DORIS_REQUEST_AUTH_PASSWORD));
        }
        return processedOptions;
    }

    private void checkOptions(Map<String, String> options) throws OptionRequiredException {
        if (!options.containsKey(DorisOptions.DORIS_FENODES.getName())) {
            throw new OptionRequiredException(DorisOptions.DORIS_FENODES.getName());
        } else {
            String feNodes = options.get(DorisOptions.DORIS_FENODES.getName());
            if (feNodes.isEmpty()) {
                throw new IllegalArgumentException("option [" + DorisOptions.DORIS_FENODES.getName() + "] is empty");
            } else if (feNodes.contains(":")) {
                if (!feNodes.matches("([\\w-.]+:\\d+,)*([\\w-.]+:\\d+)")) {
                    throw new IllegalArgumentException("option [" + DorisOptions.DORIS_FENODES.getName() + "] is not in correct format, for example: host:port[,host2:port]");
                }
            }
        }
        if (!ignoreTableCheck) {
            if (!options.containsKey(DorisOptions.DORIS_TABLE_IDENTIFIER.getName())) {
                throw new OptionRequiredException(DorisOptions.DORIS_TABLE_IDENTIFIER.getName());
            } else {
                String tableIdentifier = options.get(DorisOptions.DORIS_TABLE_IDENTIFIER.getName());
                if (tableIdentifier.isEmpty()) {
                    throw new IllegalArgumentException("option [" + DorisOptions.DORIS_TABLE_IDENTIFIER.getName() + "] is empty");
                } else if (!tableIdentifier.contains(".")) {
                    throw new IllegalArgumentException("option [" + DorisOptions.DORIS_TABLE_IDENTIFIER.getName() + "] is not in correct format, for example: db.table");
                }
            }
        }
        if (!options.containsKey(DorisOptions.DORIS_USER.getName())) {
            throw new OptionRequiredException(DorisOptions.DORIS_USER.getName());
        } else if (options.get(DorisOptions.DORIS_USER.getName()).isEmpty()) {
            throw new IllegalArgumentException("option [" + DorisOptions.DORIS_USER.getName() + "] is empty");
        }
        if (!options.containsKey(DorisOptions.DORIS_PASSWORD.getName())) {
            throw new OptionRequiredException(DorisOptions.DORIS_PASSWORD.getName());
        }
        if ("thrift".equalsIgnoreCase(options.get(DorisOptions.READ_MODE.getName()))) {
            if (Boolean.parseBoolean(options.get(DorisOptions.DORIS_READ_BITMAP_TO_STRING.getName()))) {
                throw new IllegalArgumentException(String.format("option [%s] is invalid in thrift read mode",
                        DorisOptions.DORIS_READ_BITMAP_TO_STRING.getName()));
            }
            if (Boolean.parseBoolean(options.get(DorisOptions.DORIS_READ_BITMAP_TO_BASE64.getName()))) {
                throw new IllegalArgumentException(String.format("option [%s] is invalid in thrift read mode",
                        DorisOptions.DORIS_READ_BITMAP_TO_BASE64.getName()));
            }
        } else {
            if (Boolean.parseBoolean(options.get(DorisOptions.DORIS_READ_BITMAP_TO_STRING.getName()))
                    && Boolean.parseBoolean(options.get(DorisOptions.DORIS_READ_BITMAP_TO_BASE64.getName()))) {
                throw new IllegalArgumentException("option [%s] and [%s] cannot be true at this same time");
            }
        }
    }

    public boolean contains(ConfigOption<?> option) {
        return configOptions.containsKey(option.getName());
    }

    public <T> T getValue(ConfigOption<T> option) throws OptionRequiredException {
        if (configOptions.containsKey(option.getName())) {
            String originVal = configOptions.get(option.getName());
            Class<?> typeClass = option.getTypeClass();
            if (typeClass == Integer.class) {
                return (T) typeClass.cast(Integer.parseInt(originVal));
            } else if (typeClass == Long.class) {
                return (T) typeClass.cast(Long.parseLong(originVal));
            } else if (typeClass == Double.class) {
                return (T) typeClass.cast(Double.parseDouble(originVal));
            } else if (typeClass == Boolean.class) {
                return (T) typeClass.cast(Boolean.parseBoolean(originVal));
            } else if (typeClass == String.class) {
                return (T) originVal;
            } else {
                throw new IllegalArgumentException("Unsupported config value type: " + typeClass);
            }
        } else if (option.getDefaultValue() != null) {
            return option.getDefaultValue();
        } else {
            throw new OptionRequiredException(option.getName());
        }
    }

    public void setProperty(ConfigOption<?> option, String value) throws OptionRequiredException {
        configOptions.put(option.getName(), value);
        configOptions = processOptions(configOptions);
        checkOptions(configOptions);
    }

    public Map<String, String> getSinkProperties() {
        Map<String, String> sinkProperties = new HashMap<>();
        for (Map.Entry<String, String> entry : configOptions.entrySet()) {
            if (entry.getKey().startsWith(DorisOptions.STREAM_LOAD_PROP_PREFIX)) {
                sinkProperties.put(entry.getKey().substring(DorisOptions.STREAM_LOAD_PROP_PREFIX.length()), entry.getValue());
            }
        }
        return sinkProperties;
    }

    public void merge(Map<String, String> map) throws OptionRequiredException {
        configOptions.putAll(map);
        configOptions = processOptions(configOptions);
        checkOptions(configOptions);
    }

    public Map<String, String> toMap() {
        return new HashMap<>(configOptions);
    }

    public static DorisConfig fromMap(Map<String, String> sparkConfMap, Boolean ignoreTableCheck) throws OptionRequiredException {
        return fromMap(sparkConfMap, Collections.emptyMap(), ignoreTableCheck);
    }

    public static DorisConfig fromMap(Map<String, String> sparkConfMap, Map<String, String> options, Boolean ignoreTableCheck) throws OptionRequiredException {
        Map<String, String> map = new HashMap<>(sparkConfMap);
        if (MapUtils.isNotEmpty(options)) {
            map.putAll(options);
        }
        return new DorisConfig(map, ignoreTableCheck);
    }

}